### PR TITLE
Feature/westgard rule

### PR DESCRIPTION
**Pull request by rekozzz**
New westgard rules updated in the back
Z-score and type of violations are stored in the db now
Front now has he real data coming from the db and doesn't have to calculate the z-score every time which enhance performance
NOTE: QC management page isn't working in the front so u planning to test the chart, create the qc test from the endpoints using postman or hots whatever its name is

### CODE REVIEW

File: apps/backend/src/qc-results/westgard.utils.ts
Comment (garma-a) [inline apps/backend/src/qc-results/westgard.utils.ts:1]: ### 1. Missing Rules (Not Defined/Detected)                                                                 
                                                                                                              
  The following standard and modified Westgard rules are missing from the backend logic:                      
                                                                                                              
  • **$2\text{of}3_{2s}$ Rule (Common for $N=3$):** Reject when 2 out of 3 consecutive control measurements   
  exceed the same $+2s$ or $-2s$ limit. This is highly useful when running three levels of controls (e.g., Low,
  Normal, High).                                                                                              
  • **$3_{1s}$ Rule (Common for $N=3$):** Reject when 3 consecutive control measurements exceed the same $+1s$
  or $-1s$ limit.                                                                                             
  • $6_x$ / $9_x$ Rules (Common for $N=3$): Reject when 6 or 9 consecutive control measurements fall on one   
  side of the mean.                                                                                           
  • $8_x$ / $12_x$ Rules (Common for $N=4$): Reject when 8 or 12 consecutive control measurements fall on one 
  side of the mean (often used across runs or when 4 control measurements are taken per run).                 
  • $7_T$ Rule (Trend Rule): Reject when 7 consecutive control measurements trend in the same direction (i.e.,
  get progressively higher or progressively lower). This is a vital rule for detecting gradual systematic     
  drift (e.g., reagent deterioration, instrument lamp aging, or temperature drift).  


### Summary of What is Evaluated vs. What is Missing

| Rule | Status | Type | Target Error | Notes / Discrepancy |
| :--- | :--- | :--- | :--- | :--- |
| **$1_{3s}$** | Active | Single-Lot | Random / Systematic | Correctly implemented. |
| **$1_{2s}$** | Active | Single-Lot | Warning Limit | Correctly implemented. |
| **$2_{2s}$** | Active | Single-Lot | Systematic | Only evaluates consecutive runs of the same lot; misses across-material violations in the same run. |
| **$R_{4s}$** | Incorrect | Single-Lot | Random | Evaluated between-run instead of within-run across control levels. |
| **$4_{1s}$** | Active | Single-Lot | Systematic | Evaluates only consecutive runs of the same lot; misses across-material violations. |
| $10_x$ | Active | Single-Lot | Systematic | Evaluates only consecutive runs of the same lot; misses across-material violations. |
| **$2\text{of}3_{2s}$** | Missing | Multi-Lot | Systematic | Useful for $N=3$ controls. |
| **$3_{1s}$** | Missing | Multi-Lot | Systematic | Useful for $N=3$ controls. |
| $6_x$ / $9_x$ | Missing | Multi-Lot | Systematic | Useful for $N=3$ controls. |
| $8_x$ / $12_x$ | Missing | Multi-Lot | Systematic | Useful for $N=4$ controls / multi-run. |
| $7_T$ | Missing | Single-Lot | Systematic Drift | Essential for detecting lamp/reagent aging trends. |

**Review by garma-a — CHANGES_REQUESTED** [review]
Really Great work ❤️ 
i just need some simple changes

**Comment by garma-a** [comment]
you can also check please these missing rules :heart: 
### Table of Missing & Incorrectly Implemented Rules in Your Codebase

This table details the rules that are missing or require correction in `westgard.utils.ts`.

| Rule Notation | Error Type | Official Definition | Recommended Configuration | Current Status | Required Fix / Action |
| :--- | :--- | :--- | :--- | :--- | :--- |
| **$R_{4s}$** | Random | Reject when 1 control in a run exceeds $+2s$ and another exceeds $-2s$ within-run. | $N \ge 2$ controls | Incorrectly Implemented (Evaluates between-run of a single lot) | Change logic to compare different control levels (e.g., Level 1 and Level 2) run in the same batch rather than consecutive runs of the same lot. |
| **$2_{2s}$ (Multi-Lot)** | Systematic | Reject when Level 1 and Level 2 both exceed the same $+2s$ or $-2s$ in the same run. | $N \ge 2$ controls | Missing (Only checks single lot consecutively) | Fetch current run values of all active control lots for a test and check if more than one exceeds the same $2s$ threshold in the same batch. |
| **$4_{1s}$ (Multi-Lot)** | Systematic | Reject when 4 consecutive measurements exceed the same $+1s$ or $-1s$ across control levels (e.g., 2 consecutive points on Level 1 + 2 consecutive points on Level 2). | $N \ge 2$ controls | Missing (Only checks single lot consecutively) | Evaluate the last 4 runs across all control levels belonging to the same test. |
| $10_x$ (Multi-Lot) | Systematic | Reject when 10 consecutive measurements fall on the same side of the mean across control levels (e.g., 5 consecutive points on Level 1 + 5 consecutive points on Level 2). | $N \ge 2$ controls | Missing (Only checks single lot consecutively) | Evaluate the last 10 runs across all control levels belonging to the same test. |
| **$2\text{of}3_{2s}$** | Systematic | Reject when 2 out of 3 consecutive measurements exceed the same $+2s$ or $-2s$ limit. | $N = 3$ controls (or $N=1$ across 3 runs) | Missing | Add helper logic to check if any 2 of the last 3 z-scores are both $> +2$ or both $< -2$. |
| **$3_{1s}$** | Systematic | Reject when 3 consecutive measurements exceed the same $+1s$ or $-1s$ limit. | $N = 3$ controls (or $N=1$ across 3 runs) | Missing | Check if the last 3 z-scores are all $\ge 1$ or all $\le -1$. |
| $6_x$ / $9_x$ | Systematic | Reject when 6 or 9 consecutive measurements fall on one side of the mean. | $N = 3$ controls (or across multiple runs) | Missing | Implement similar logic to $10_x$, but checking for a window size of 6 or 9. |
| $8_x$ / $12_x$ | Systematic | Reject when 8 or 12 consecutive measurements fall on one side of the mean. | $N = 2$ or $N = 4$ controls | Missing | Add configurations to check window sizes of 8 and 12 on one side of the mean. |
| $7_T$ (Trend) | Systematic | Reject when 7 consecutive measurements get progressively higher or progressively lower. | All $N$ configurations | Missing | Check if the last 7 z-scores are strictly monotonic: $z_0 > z_1 > z_2 \dots > z_6$ or $z_0 < z_1 < z_2 \dots < z_6$. |

@rekozzz

**Comment by rekozzz** [comment]
Fixed few rules but didn't do the multi lots things yet due some changes will needed to be discussed first .
fixed the refresh thing when u refresh the page when u are in the analytics page it get u back to overview and i put cursor pointy thingy as well.

**Comment by garma-a** [comment]
### 🔴 HIGH — Frontend status string mismatches (UI never shows correct state)                              
                                                                                                              
  File: QCHistory.tsx                                                                                        
                                                                                                              
   Line │ Bug                                       │ Code                        │ Should Be
  ──────┼───────────────────────────────────────────┼─────────────────────────────┼───────────────────────────
   L105 │ Severity never set to  'reject'           │  t.status === 'error'       │  t.status === 'FAIL' 
   L119 │ Status never becomes  'reject'            │  t.status === 'error'       │  t.status === 'FAIL' 
   L356 │ CheckCircle never renders (PASS icon      │  qc.status === 'pass'       │  qc.status === 'PASS' 
        │ missing)                                  │                             │
   L312 │ Badge shows on ALL points including PASS  │  point.status !== 'normal'  │  point.status !== 'PASS' 
                                                                                                              
  Backend sends  'PASS'  /  'FAIL'  /  'WARNING'  (uppercase). The frontend compares against  'error' ,       
  'pass' , and  'normal'  — none of which ever match.                                                         
  ──────                                                                                                      
  ### 🟡 MEDIUM —  findOne  still guards on lot stats unnecessarily                                           
                                                                                                              
  File: qc-results.service.ts Lines 176–183                                                                         
                                                                                                              
    if (                                                                                                      
      result.control_lots!.mean === null ||         // line 177                                               
      result.control_lots!.standardDeviation === null  // line 178                                            
    ) {                                                                                                       
      throw new BadRequestException(...)             // line 180                                              
    }                                                                                                         
                                                                                                              
  Z-scores are now stored on  qc_results  at submission time. This guard blocks reads of valid historical     
  results if a control lot's stats are ever cleared/updated.                                                  
  ──────                                                                                                      
  ### 🟡 MEDIUM —  1₃s  boundary uses  >= 3  instead of  > 3                                                  
  File: westgard.utils.ts Line 111
  
    if (Math.abs(z) >= 3) {   // triggers at z = exactly ±3.0
    
  Standard Westgard: a point exactly at ±3SD is a warning, not a reject. Should be  > 3 .
  ──────
  ### 🟡 MEDIUM — Alert result matching by array index (fragile)
  
  File: qc-results.service.ts Lines 118–120
  
    for (let i = 0; i < evaluatedResults.length; i++) {
      const savedResult = savedRunData.results[i];  // line 120 — index-based match
    
   INSERT … RETURNING  order is not contractually guaranteed by Postgres. Should map by  lotId  instead.
  ──────                                                                                                      
  ### 🟡 MEDIUM —  getResultsByLotId  still missing  performedBy 
  
  File: qc-results.repository.ts Lines 98–116
  
  The select at line 100–110 includes  runId  but not  performedBy . Since  performedBy  moved to  qcRuns , it
  should select  qcRuns.performedBy  — and  QcResultResponseDto  (line 42 in the DTO file) still declares     
  performedBy: number  as required

**Comment by garma-a** [comment]
### ⚠️ This PR will overwrite the React Query infrastructure added in  main  (commit  203f737 )            
  │                                                                                                           
  │ The following files exist on  main  and are not present in this branch at all. Merging this PR will delete
  │ them:                                                                                                     
  │                                                                                                           
  │  File                                            │ What it does                                           
  │                                                                                                           
  ─────────────────────────────────────────────────┼───────────────────────────────────────────────────────── 
  │   apps/frontend/src/lib/query/QueryProvider.tsx  │ App-wide  QueryClient  with 30s stale time, 5min GC,   
  │                                                  │ refetchOnWindowFocus  — wraps the entire app in        
  │                                                  │ layout.tsx                                             
  │   apps/frontend/src/lib/api/clientFetch.ts       │ Client-side fetch helper that forwards  AbortSignal    
  │                                                  │ from React Query to cancel in-flight requests on       
  │                                                  │ component unmount / lot switch                         
  │   apps/frontend/src/hooks/useQcResults.ts        │  useQuery  hook for QC results — AbortController-      
  │                                                  │ backed, per-lot cache key, cancels stale requests when 
  │                                                  │ user switches lots rapidly                             
  │   apps/frontend/src/hooks/useAlerts.ts           │  useQuery  +  useMutation  for alerts — includes       
  │                                                  │ pollIntervalMs  for live polling                       
  │   apps/frontend/src/hooks/useMachines.ts         │  useQuery  hook for machines                           
  │   apps/frontend/src/hooks/useControlLots.ts      │  useQuery  hook for control lots                       
  │   apps/frontend/src/hooks/useQcTests.ts          │  useQuery  hook for QC tests                           
  │                                                                                                           
  │ Additionally,  apps/frontend/package.json  on  main  has these dependencies that are missing from this    
  │ branch's  package.json :                                                                                  
  │                                                                                                           
  │   "@tanstack/react-query": "^5.100.14",                                                                   
  │   "@tanstack/react-query-devtools": "^5.100.14"                                                           
  │                                                                                                           
  │ Merging will remove these from the lockfile.                                                              
  │                                                                                                           
  │ And  apps/frontend/src/app/layout.tsx  on  main  wraps the app with  <QueryProvider>  — this branch's     
  │ version of the layout does not include it, so the  QueryProvider  tree will be gone even if the files     
  │ survive.                                                                                                  
  │                                                                                                           
  │ The following frontend files are modified by this PR AND exist on  main  with React Query usage patterns  
  │ that will be lost:
  │
  │  File in this PR                                     │ Impact
  │
  ─────────────────────────────────────────────────────┼───────────────────────────────────────────────────── 
  │   apps/frontend/src/app/qc/page.tsx 		│ Replaced with a plain async Server Component — no  
  │                                                      │ client-side refetch or cache
  │   apps/frontend/src/app/monitor/page.tsx             │ Same — server-only one-shot fetch
  │   apps/frontend/src/features/machines/components/Moni│ Receives data only as props, no  useQuery          
  │  torClient.tsx                                       │
  │   apps/frontend/src/features/machines/components/Mach│ Same
  │  ineCharts.tsx                                       │
  │   apps/frontend/src/features/qc/components/QCHistoryI│ Same
  │  nteractive.tsx                                      │
  │   apps/frontend/src/features/qc/components/QCHistory.│ Same
  │  tsx                                                 │
  │
  │ Requested action before merge:
  │ Please rebase this branch on top of  main  ( git rebase main ) so the React Query infrastructure is       
  │ preserved, and update the modified pages/components to use the existing hooks ( useQcResults ,            
  useMachines
  │ , etc.) instead of server-side  await api.get()  calls. The backend changes in this PR are solid — the    
  │ conflict is purely in the frontend data-fetching layer.
  ──────
  That's the full comment — covers every deleted file, the missing package, the layout regression, and tells  
  the author exactly how to fix it.

**Comment by garma-a** [comment]
@rekozzz

File: apps/backend/src/drizzle/schema.ts [hunk 1]
```diff
@@ -97,25 +97,39 @@ export const controlLots = pgTable('control_lots', {
...
+  id: serial('id').primaryKey(),
+  machineId: integer('machine_id')
+    .references(() => machines.id)
+    .notNull(),
+  performedBy: integer('performed_by')
+    .references(() => users.id)
+    .notNull(),
+  runDate: timestamp('run_date').defaultNow(),
+});
+
```
Comment (garma-a) [inline apps/backend/src/drizzle/schema.ts:119]: qc_run  points to a  machine , but NOT to a  qc_test . So the relationship chain is:                       
                                                                                                              
    qc_run → machine ✅                                                                                       
    qc_result → control_lot → qc_test → machine ✅                                                            
                                                                                                              
    But:                                                                                                      
    qc_run has NO direct reference to qc_test ⚠️                                                               
                                                                                                              
  This means a single run could theoretically mix results from different QC tests on the same machine (e.g.,  
  HGB Level 1 + Platelet Level 2 in the same run), which is clinically incorrect — a run is always for one    
  test at a time.                                                                                             
                                                                                                              
  The fix would be adding  testId  to  qc_runs :                                                              
                                                                                                              
    qcRuns = pgTable('qc_runs', {                                                                             
      ...                                                                                                     
      testId: integer('test_id').references(() => qcTests.id).notNull(), // missing!                          
      ...                                                                                                     
    })

File: apps/backend/src/qc-results/qc-results.service.ts [hunk 2]
```diff
@@ -10,6 +10,19 @@ import { QcStatus } from './qc-results.types';
 import { AlertsService } from '@/alerts/alerts.service';
 import { AlertPriority } from '@/alerts/alerts.types';
 import { UsersRepository } from '@/users/users.repository';
+import { evaluateWestgardRules, WESTGARD_HISTORY_SIZE, RunResultItem } from './westgard.utils';
+import { QcResultItemDto } from './dto/create-qc-result.dto';
```
Comment (garma-a) [inline apps/backend/src/qc-results/qc-results.service.ts:14]: please use '@/dto/create-qc-result.dto'; instead of './dto/create-qc-result.dto';
so the @/ instead of the ./

**Review by garma-a — CHANGES_REQUESTED** [review]

File: apps/backend/src/qc-results/westgard.utils.ts
Comment (garma-a) [inline apps/backend/src/qc-results/westgard.utils.ts:1]: ### Missing Dedicated Unit Tests for Westgard Math

* **Integration Tests:** `qc-results.service.spec.ts` covers the service layer very well. It correctly tests that a R₄ₛ cross-material violation fails both levels correctly.
* **Missing Edge Cases:** There is no dedicated `westgard.utils.spec.ts` file. `westgard.utils.ts` contains dense, pure-math functions evaluating 10+ clinical rules.

> [!IMPORTANT]  
> **Action Required:** You must create a `westgard.utils.spec.ts` to exhaustively test edge cases, including:
> 
> * **Boundary conditions:** e.g., what happens when `Math.abs(z) === 3` vs `> 3`. Currently, the code uses `> 3` for 1₃ₛ, meaning exactly 3.0 passes the rule. Is this intentional per clinical guidelines, or should it be `>= 3`?
> * **Rule 7T:** Check that exactly 7 consecutive ascending/descending points trigger the trend violation, but 6 do not.
> * **Empty Arrays:** Passing empty history or missing lot IDs.

**Review by garma-a — CHANGES_REQUESTED** [review]

**Comment by rekozzz** [comment]
please merge it, no more fixing. i wanna move on and work on something new ❤️😭
